solutions/GCJ-BacterialTactics.mdx

```diff
@@ -0,0 +1,118 @@
+---
+id: GCJ-BacterialTactics
+source: Round 1C 2019 - Code Jam 2019
+title: Bacterial Tactics
+author: Shijie Ren
+---
+
+[Offcial Solution](https://codingcompetitions.withgoogle.com/codejam/round/00000000000516b9/0000000000134cdf#analysis)
+
+## Explanation
+
+First, take a look of the game: The player delete a column or a line without
+radioactive material, and divide the matrix into two small games, still
+following the rules above, the one who cannot operate will lose.
+
+Since the matrix is small here, it is appropriate to calculate SG for every
+submatrix. According to the Sprague–Grundy Theorem, the SG of two subgames will
+be the bitwise XOR sum of them.
+
+Let $f[i][j][a][b]$ be the SG value of submatrix where $(i,j)$ in the upper left
+corner and $(a,b)$ in the lower right corner. Thus,
+
+$f[i][j][a][b]=\text{MEX}(f[i][p-1][a][b] \oplus f[p+1][j][a][b],f[i][j][a][q-1] \oplus f[i][j][q+1][b])$
+
+Where $\oplus$ reprents bitwise XOR operation. Notice that there must not be
+radioactive material on the chosen line or column, this can be pre-calculated in
+$\mathcal{O}(N^2)$, for each cell pre-calculate the position of next cell with
+radioactive material in a specific direction.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^5)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define N 17
+char s[N][N];
+int f[N][N][N][N];
+int vis[N][N][N][N];
+int nex1[N][N];
+int nex2[N][N];
+int n, m, ans = 0;
+int work(int x, int y, int l, int r) {
+	if (x > y || l > r)
+		return 0;
+	if (vis[x][y][l][r])
+		return f[x][y][l][r];
+	int p, sum = 0, tmp = 0;
+	map<int, int> s;
+	for (p = x; p <= y; p++) {
+		tmp = work(x, p - 1, l, r) ^ work(p + 1, y, l, r);
```

Comment:
comment your code, walk us through what this is doing
for reference you can look at https://github.com/cpinitiative/usaco-guide/pull/2994